### ISSUE TITLE
Crash if unsupported language code used

### ISSUE THREAD

**Issue by timj** [ecdye/macSubtitleOCR#36]
[crash.sup.gz](https://github.com/user-attachments/files/17820591/crash.sup.gz)

These are Arabic subtitles, which the API doesn't support but which fails in a terrifying way.

```
$ macSubtitleOCR --languages ar crash.sup
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString _isNaturallyRTL]: unrecognized selector sent to instance 0x9dfb1da9d10d90d1'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000191ad4300 __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x00000001915bacd8 objc_exception_throw + 88
	2   CoreFoundation                      0x0000000191b89638 -[NSObject(NSObject) __retain_OA] + 0
	3   CoreFoundation                      0x0000000191a40cc0 ___forwarding___ + 1568
	4   CoreFoundation                      0x0000000191a405e0 _CF_forwarding_prep_0 + 96
	5   TextRecognition                     0x00000001c1efa0a8 _ZN20CRDetectorOutputMaps13aggregateMapsEPK27CRTextDetectorModelV3OutputRK6CGSizeRK7CGPoint + 53144
	6   TextRecognition                     0x00000001c1e1861c CROSLogForCategory + 117252
	7   TextRecognition                     0x00000001c1e10ea0 CROSLogForCategory + 86664
	8   TextRecognition                     0x00000001c1e0dca4 CROSLogForCategory + 73868
	9   TextRecognition                     0x00000001c1e88184 CRSignpostLog + 5556
	10  TextRecognition                     0x00000001c1f184a8 _ZN15CHNeuralNetwork11FeedForwardEPfii + 63016
	11  TextRecognition                     0x00000001c1e87f38 CRSignpostLog + 4968
	12  libdispatch.dylib                   0x00000001917c68f8 _dispatch_call_block_and_release + 32
	13  libdispatch.dylib                   0x00000001917c8658 _dispatch_client_callout + 20
	14  libdispatch.dylib                   0x00000001917cbb8c _dispatch_continuation_pop + 728
	15  libdispatch.dylib                   0x00000001917cb138 _dispatch_async_redirect_invoke + 580
	16  libdispatch.dylib                   0x00000001917da314 _dispatch_root_queue_drain + 392
	17  libdispatch.dylib                   0x00000001917dab20 _dispatch_worker_thread2 + 156
	18  libsystem_pthread.dylib             0x000000019197739c _pthread_wqthread + 228
	19  libsystem_pthread.dylib             0x00000001919760f0 start_wqthread + 8
)
libc++abi: terminating due to uncaught exception of type NSException
```

**Comment by ecdye** [comment]
Fun! Looks like a fun one, I'll take a look. That's a good edge case.

**Comment by ecdye** [comment]
I don't have a resolution yet, but interestingly it appears that Arabic is supported, or at least it appears in the list of supported languages when I print them while debugging. I'll have to do some digging, but a small part of me wonders if this might be an issue with the Vision API

**Comment by ecdye** [comment]
OK, so I am pretty sure that it is an issue with the Vision API, I did some more debugging and it crashes even when I tell it to automatically detect the language. I'm not sure how to resolve the error, as I can't even catch it. Any suggestions? I could just add a note, or try and check if the user is trying to specify arabic and print an error and exit, but I don't particularly care for hardcoding a case like that.

**Comment by timj** [comment]
I was wondering if it was a right-to-left language thing but `he` (Hebrew) works fine.

Of course Apple software can handle Arabic just fine (there would be many complaints if it couldn't) if I copy and paste from a screen grab.

If I turn on `--save-images` it does write some images before the crash (and with `--invert` a different set of images). 1 to 4 exist but then it skips to 20 or 81 -- this seems confusing.

**Comment by timj** [comment]
And if I use `-l he` it saves far more images

**Comment by ecdye** [comment]
Yeah, thats what is throwing me off too, it seems like it is a bug in the API and I have no good way to prevent the hard crash. I suspect the seemingly random images being output is just because of the multithreading queueing all of the different images and then choosing (I assume at random) the next image to process when a thread is free.

**Comment by ecdye** [comment]
@timj Good news, I finally heard back from Apple on my bug report about this. They told me that it should be resolved in the next macOS release (15.3). Hopefully that is the case, I'll try and test it over the next week or two and assuming it works I'll close this issue.

**Comment by timj** [comment]
Thanks for filing a ticket in this. Great that they knew about it and are fixing it.

**Comment by ecdye** [comment]
I can confirm that the fix appears to have fully resolved the issue, when 15.3 is released it should work well.
